pkg/mutation/schema/node.go

```diff
@@ -0,0 +1,241 @@
+	if len(path) == 0 {
+		return nil
+	}
+
+	// Initialize child within n.children.
+	childNode := path[0]
+	if n.Children == nil {
+		n.Children = make(map[string]map[parser.NodeType]node)
+	}
+	childKey := childNode.Key()
+	if n.Children[childKey] == nil {
+		n.Children[childKey] = make(map[parser.NodeType]node)
+	}
+	childType := headType(path)
+	if _, exists := n.Children[childKey][childType]; !exists {
+		n.Children[childKey][childType] = node{}
+	}
+
+	// Add the remaining path to the appropriate child, collecting any conflicts
+	// found when adding it.
+	child := n.Children[childKey][childType]
+	conflicts := child.Add(id, path[1:])
+	n.Children[childKey][childType] = child
+
+	// Detect conflicts at this node.
+	// We know there is a conflict if there is a child with the same Key but a
+	// different type.
+	conflicts = merge(conflicts, n.conflicts(childKey))
+	return conflicts
+}
+
+const ErrNotFound = util.Error("path not found")
+
+func (n *node) Remove(id types.ID, path []parser.Node) error {
+	// This ID no longer references this node.
+	if _, isReferenced := n.ReferencedBy[id]; isReferenced {
+		delete(n.ReferencedBy, id)
+	} else {
+		return ErrNotFound
+	}
+
+	if len(path) == 0 {
+		// No more path to remove.
+		return nil
+	}
+
+	childKey := path[0].Key()
+	if _, found := n.Children[childKey]; !found {
+		// The child does not exist.
+		return fmt.Errorf("no key %q: %w", childKey, ErrNotFound)
+	}
+	childType := headType(path)
+	if _, found := n.Children[childKey][childType]; !found {
+		// No child of the key and type exists.
+		return fmt.Errorf("no type %q for key %q: %w", childType, childKey, ErrNotFound)
+	}
+
+	child := n.Children[childKey][childType]
+	err := child.Remove(id, path[1:])
+	if err != nil {
+		return fmt.Errorf("%q: %w", childKey, err)
+	}
+
+	// Delete the type from the child if it is no longer referenced.
+	if len(child.ReferencedBy) == 0 {
+		// No references to this child of this type exist.
+		delete(n.Children[childKey], childType)
+	} else {
+		n.Children[childKey][childType] = child
+	}
+
+	// Delete the child if it is no longer referenced.
+	if len(n.Children[childKey]) == 0 {
+		delete(n.Children, childKey)
+	}
+
+	return nil
+}
+
+func (n *node) conflicts(childKey string) idSet {
+	conflicts := make(idSet)
+
+	// Count the number of distinct types with this key.
+	nTypes := len(n.Children[childKey])
+	if _, hasUnknown := n.Children[childKey][unknown]; hasUnknown {
+		// Nodes whose types we are unable to determine do not count against this
+		// check.
+		nTypes--
+	}
+
+	for nodeType, child := range n.Children[childKey] {
+		if nodeType == unknown {
+			// If we don't know the type of a node, we assume it conflicts with nothing.
+			continue
+		}
+		// There are conflicts if either:
+		// 1) there are more than one non-unknown types for the Child, or
+		// 2) the Child is a List and defines multiple keys.
+		if nTypes > 1 || nodeType == parser.ListNode && len(child.Children) > 1 {
```

Comment:
This qualifier for how to interpret `len(child.Children)` shows how we are pulling double-duty on how the `Children` map is defined.

Because this is a purely internal implementation, I'm fine with this.

Reply:
I couldn't find a simple alternative that didn't repeat logic.

Reply:
The alternative would be to have two different types of nodes, one for list one for objects Not saying it's a better approach. IMO it's probably a wash, with different types being preferable if we thought this would get more complicated (I don't)